### PR TITLE
compact param in calls

### PR DESCRIPTION
**Pull request by thiolliere**
related to https://github.com/paritytech/substrate/issues/907

In order to replace `<T as HasCompact>::Type` in call parameter we introduce a new attribute that allow Call enum to derive Encode and Decode with compact parameters. for instance on timestamp:
```rust
-		fn set(origin, now: <T::Moment as HasCompact>::Type) {
+		fn set(origin, #[compact] now: T::Moment) {
```

This PR also rewrite evey such usage as such.

Also one consequence is that lots of `.into()` aren't required anymore.

This also makes me wonder about the future of Compact struct: Do every usage of `Compact<not generic type>` should be replaced the same way ? it seems so to me as it is more practical.

Finally I don't know if this have implications beyond this codebase, I only tested with `cargo test --all` and eveything seems fine inside this codebase.

Syntax could be very easily changed, it was made this way so it is quite coherent with attribute when deriving codec for structs and enum.

### CODE REVIEW

File: srml/support/src/dispatch.rs [hunk 1]
```diff
@@ -481,6 +481,8 @@ macro_rules! decl_module {
 				$origin_type;
 				$from;
 				$fn_vis fn $fn_name (
+					// TODO: do we allow those meta here probbly not but so we have to enforce that
```
Comment (bkchr) [inline srml/support/src/dispatch.rs]: Every `TODO` should be an issue in the github issue tracker and the issue link should be attached to the comment.

(Besides that, I don't understand this `TODO`)

File: srml/support/src/dispatch.rs [hunk 2]
```diff
@@ -571,17 +573,18 @@ macro_rules! decl_module {
...
 
 		impl<$trait_instance: $trait_name> $crate::dispatch::Encode for $call_type<$trait_instance> {
 			fn encode_to<W: $crate::dispatch::Output>(&self, _dest: &mut W) {
-				__impl_encode!(_dest; *self; 0; $call_type; $( fn $fn_name( $( $param_name ),* ); )*);
+				__impl_encode!(_dest; *self; 0; $call_type; $( fn $fn_name( $( $(#[$codec_attr on type $param])* $param_name ),* ); )*);
 				if let $call_type::__PhantomItem(_) = *self { unreachable!() }
 				if let $call_type::__OtherPhantomItem(_) = *self { unreachable!() }
 			}
 		}
+		// TODO: do we need to change this also ?
```
Comment (bkchr) [inline srml/support/src/dispatch.rs]: I think we don't need this. The important pieces are the `Encode` and `Decode` impls.

File: srml/support/src/dispatch.rs [hunk 3]
```diff
@@ -608,11 +611,13 @@ macro_rules! decl_module {
 			type Call = $call_type<$trait_instance>;
 		}
 
+		// TODO: seems ok but not sure
```
Comment (bkchr) [inline srml/support/src/dispatch.rs]: Same.

File: srml/support/src/dispatch.rs [hunk 4]
```diff
@@ -651,7 +660,27 @@ macro_rules! __impl_decode {
...
+		$input:expr;
+	) => {
+		let $param_name = $crate::dispatch::Decode::decode($input)?;
+	};
+	(@decode
+		$(#[$codec_attr:ident on type])*
+		$param_name:ident;
+		$input:expr;
+	) => {
+		compile_error!(concat!("Invalid attribute for parameter `", stringify!($param_name), "`, should be one attribute `#[compact]`"))
```
Comment (bkchr) [inline srml/support/src/dispatch.rs]:
```suggestion
		compile_error!(concat!("Invalid attribute for parameter `", stringify!($param_name), "`, the following attributes are supported: `#[compact]`"))

```

File: srml/support/src/dispatch.rs [hunk 5]
```diff
@@ -687,7 +720,27 @@ macro_rules! __impl_encode {
...
+		$dest:expr;
+	) => {
+		$param_name.encode_to($dest);
+	};
+	(@encode_as
+		$(#[$codec_attr:ident on type $param:ty])*
+		$param_name:ident;
+		$dest:expr;
+	) => {
+		compile_error!(concat!("Invalid attribute for parameter `", stringify!($param_name), "`, should be one attribute `#[compact]`"))
```
Comment (bkchr) [inline srml/support/src/dispatch.rs]:
```suggestion
		compile_error!(concat!("Invalid attribute for parameter `", stringify!($param_name), "`, the following attributes are supported: `#[compact]`"))

```

**Comment by bkchr** [comment]
@cheme this requires support for by metadata as well.
@jacogr any preferences on how to put that into metadata?

**Comment by jacogr** [comment]
No preferences, we cater for (using `Balance` as an example) `Compact<Balance>` and `<Balance as HasCompact>` when we do the parsing. 

If at all, the first form, e.g. `Compact<Balance>` is easier all-around, is slightly more compact and seems less like the implementation details bleeds through into the metadata.

**Comment by gavofyork** [comment]
Agree `Compact<Whatever>` would be easiest.

And would be cool to switch out `x: Compact<u64>` ... `x.into()` for `#[compact] x: u64` ... `x`

File: srml/support/src/dispatch.rs (hunk 1 above)
Comment (thiolliere) [reply srml/support/src/dispatch.rs]: (it was a personnal as other TODO below reminder, sorry)
it actually means that if there was other attribute #[..] in front of a parameter they wouldn't be used in code generation apart from our compact use.
But as it wasn't allowed before and we don't allow any other things. It is completely OK

**Comment by gavofyork** [comment]
Yeah... right now compact arguments are completely unmarked in the metadata: they'll need to be surrounded with `Compact<...>`.

**Comment by thiolliere** [comment]
Yes I work on metadata

**Comment by thiolliere** [comment]
* test checking `#[compact]` is actually compact implemented
* metadata implemented

**Pull request merged**